### ISSUE TITLE
[BUG] Unable to uninstall Longhorn via Helm

### ISSUE THREAD

**Issue by githubixx** [longhorn/longhorn#5227]
## Describe the bug

I'm currently playing around with Longhorn before doing more serious stuff. So I created a K8s test cluster with 3 nodes by using Vagrant and Linux KVM. I installed Longhorn via Helm as described in the Longhorn documentation without changing any default values (for now):

```
helm install longhorn longhorn/longhorn --namespace longhorn-system --create-namespace --version 1.4.0
```

Installation worked fine and I was able to create a volume. Then I deleted that volume again. Since I wanted to try out different settings I wanted to delete all the Longhorn resources to be able to start again. So I just used

```
helm uninstall longhorn -n longhorn-system
```

as stated in the documentation. AFAICS everything got deleted but deleting `longhorn-system` namespace stuck. `kubectl get ns longhorn-system` shows:

```
NAME              STATUS        AGE
longhorn-system   Terminating   11h
```

If I do `kubectl get ns longhorn-system -o yaml` I get this:

```
  Type                                         Status  LastTransitionTime               Reason                Message
  ----                                         ------  ------------------               ------                -------
  NamespaceDeletionDiscoveryFailure            False   Fri, 06 Jan 2023 19:23:42 +0000  ResourcesDiscovered   All resources successfully discovered
  NamespaceDeletionGroupVersionParsingFailure  False   Fri, 06 Jan 2023 19:23:42 +0000  ParsedGroupVersions   All legacy kube types successfully parsed
  NamespaceDeletionContentFailure              False   Fri, 06 Jan 2023 20:10:13 +0000  ContentDeleted        All content successfully deleted, may be waiting on finalization
  NamespaceContentRemaining                    True    Fri, 06 Jan 2023 20:10:13 +0000  SomeResourcesRemain   Some resources are remaining: engineimages.longhorn.io has 1 resource instances, nodes.longhorn.io has 3 resource instances
  NamespaceFinalizersRemaining                 True    Fri, 06 Jan 2023 20:10:13 +0000  SomeFinalizersRemain  Some content in the namespace has finalizers remaining: longhorn.io in 4 resource instances
```

So listing all `engineimages.longhorn.io` resources with `kubectl -n longhorn-system get engineimages.longhorn.io ` shows this:

```
NAME          STATE      IMAGE                               REFCOUNT   BUILDDATE   AGE
ei-fc06c6fb   deployed   longhornio/longhorn-engine:v1.4.1   0          7d21h       11h
```

And `kubectl -n longhorn-system get nodes.longhorn.io -A` shows this:
```
NAMESPACE         NAME             READY   ALLOWSCHEDULING   SCHEDULABLE   AGE
longhorn-system   test-longhorn1   True    true              True          11h
longhorn-system   test-longhorn2   True    true              True          11h
longhorn-system   test-longhorn3   True    true              True          11h
```

Trying to delete any of the above resources doesn't work. E.g. `kubectl -n longhorn-system delete engineimages.longhorn.io ei-fc06c6fb` just outputs `engineimage.longhorn.io "ei-fc06c6fb" deleted` but it stuck there. You have to "CTRL+C" as this command never comes back. Same for `kubectl -n longhorn-system delete nodes.longhorn.io ...`.

When I do `journalctl -f` on one of the nodes I get this error constantly:

```
Jan 06 20:17:37 test-longhorn1 kube-apiserver[11549]: W0106 20:17:37.716967   11549 reflector.go:424] storage/cacher.go:/longhorn.io/nodes: failed to list longhorn.io/v1beta1, Kind=Node: conversion webhook for longhorn.io/v1beta2, Kind=Node failed: Post "https://longhorn-conversion-webhook.longhorn-system.svc:9443/v1/webhook/conversion?timeout=30s": service "longhorn-conversion-webhook" not found
Jan 06 20:17:37 test-longhorn1 kube-apiserver[11549]: E0106 20:17:37.717003   11549 cacher.go:440] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list longhorn.io/v1beta1, Kind=Node: conversion webhook for longhorn.io/v1beta2, Kind=Node failed: Post "https://longhorn-conversion-webhook.longhorn-system.svc:9443/v1/webhook/conversion?timeout=30s": service "longhorn-conversion-webhook" not found; reinitializing...
```

But somehow this is obvious as `helm uninstall` most probably will also delete the webhooks. So what I did too was to delete some webhooks still left e.g.:

```
kubectl -n longhorn-system delete mutatingwebhookconfigurations.admissionregistration.k8s.io longhorn-webhook-mutator

kubectl -n longhorn-system delete validatingwebhookconfigurations.admissionregistration.k8s.io longhorn-webhook-validator
```

What's also still left and what I'm not able to delete are some "apiservices" e.g.:

```
kubectl get apiservices.apiregistration.k8s.io  | grep longhorn
v1beta1.longhorn.io                    Local     True        53m
v1beta2.longhorn.io                    Local     True        49m
```

Same for `CRD`s:

```
kubectl get crds -A | grep longhorn
backingimagedatasources.longhorn.io          2023-01-06T09:13:04Z
backingimagemanagers.longhorn.io             2023-01-06T09:13:04Z
backingimages.longhorn.io                    2023-01-06T09:13:04Z
backups.longhorn.io                          2023-01-06T09:13:04Z
backuptargets.longhorn.io                    2023-01-06T09:13:04Z
backupvolumes.longhorn.io                    2023-01-06T09:13:04Z
engineimages.longhorn.io                     2023-01-06T09:13:04Z
nodes.longhorn.io                            2023-01-06T09:13:04Z
```

As mentioned above Longhorn pods are at least all gone. So only Cilium and CoreDNS is left:

```
kubectl get pods -A -o wide
NAMESPACE     NAME                               READY   STATUS    RESTARTS      AGE    IP              NODE             NOMINATED NODE   READINESS GATES
cilium        cilium-888hs                       1/1     Running   0             2d9h   192.168.10.20   test-longhorn2   <none>           <none>
cilium        cilium-bbmdh                       1/1     Running   0             2d9h   192.168.10.10   test-longhorn1   <none>           <none>
cilium        cilium-knqlg                       1/1     Running   1 (54m ago)   2d9h   192.168.10.30   test-longhorn3   <none>           <none>
cilium        cilium-operator-76c6d95785-vbggb   1/1     Running   0             2d9h   192.168.10.10   test-longhorn1   <none>           <none>
cilium        cilium-operator-76c6d95785-z424z   1/1     Running   0             2d9h   192.168.10.20   test-longhorn2   <none>           <none>
kube-system   coredns-944845b57-jfv8m            1/1     Running   0             49m    10.0.0.60       test-longhorn2   <none>           <none>
kube-system   coredns-944845b57-q4mzb            1/1     Running   0             11h    10.0.2.31       test-longhorn1   <none>           <none>
```

Is there a way to fix this or do I really have to delete the whole cluster? I'm mean for the test setup that's not so much of an issue but if that happens in a more production like environment then it's not so funny :wink: 

## To Reproduce

See above.

## Expected behavior

`helm uninstall` should just delete all Longhorn resources and should not result in a `longhorn-system` namespace that stays in `Terminating` state forever.

## Log or Support bundle

As most Longhorn resources are gone after `helm uninstall ...` this is not possible.

## Environment

 - Longhorn version: v1.4.0
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: Kubernetes v1.25.5 installed via Ansible
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 3 (on same hosts as the management nodes)
 - Node config
   - OS type and version: 20.04.5 LTS (Focal Fossa) 
   - CPU per node: 2
   - Memory per node: 4 GB
   - Disk type(e.g. SSD/NVMe): SSD
   - Network bandwidth between the nodes: 3 VMs on the same host (just a test system)
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal
 - Number of Longhorn volumes in the cluster: 0 as I was about to uninstall Longhorn.

**Comment by khushboo-rancher** [comment]
@githubixx Did you set the `deleting-confirmation-flag` to true before trying to uninstall Longhorn?

https://longhorn.io/docs/1.4.0/deploy/uninstall/#prerequisite

<img width="818" alt="Screenshot 2023-01-06 at 3 36 28 PM" src="https://user-images.githubusercontent.com/60111667/211117580-4d872622-b9c2-4775-962a-20b39e2dcd02.png">

**Comment by githubixx** [comment]
@khushboo-rancher Sure. Otherwise deletion isn't possible. I forgot to set it the first time but then `helm uninstall ...` returns with an error AFAIR and the only thing that happens is that you get two additional pods called something like `longhorn-uninstall...`. As already mentioned all resources are gone besides the four mentioned above: 3x `nodes.longhorn.io` and 1x `engineimages.longhorn.io`.

**Comment by githubixx** [comment]
I did a `kubectl -n longhorn-system describe settings.longhorn.io` just before I deleted Longhorn via `helm uninstall ...`. I piped the output in a file. Here is the relevant part (besides other settings of course):

```
...

Name:         deleting-confirmation-flag
Namespace:    longhorn-system
Labels:       <none>
Annotations:  longhorn.io/configmap-resource-version: 413867
API Version:  longhorn.io/v1beta2
Kind:         Setting
Metadata:
  Creation Timestamp:  2023-01-06T09:36:52Z
  Generation:          1
  Managed Fields:
    API Version:  longhorn.io/v1beta2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:longhorn.io/configmap-resource-version:
      f:value:
    Manager:         longhorn-manager
    Operation:       Update
    Time:            2023-01-06T09:36:52Z
  Resource Version:  417577
  UID:               d6f4f3a5-403b-4baf-893b-e1abb2d5ccfd
Value:               true
Events:              <none>

....
```

**Comment by githubixx** [comment]
Ok, I found the problem. It's always this webhook thingy if the controller nodes don't have any idea of ClusterIPs. My controller and worker nodes are on the same network in general. No restrictions there. But there is no `kube-proxy` running on the controller nodes. So they are not able to connect to ClusterIPs and therefore webhooks are timing out as they are initiated by the nodes which has `kube-apiserver` process running and that's normally the controller nodes.

So if someone comes across this issue: I also installed `kube-proxy` and `kubelet` now on the controller nodes but added taints so that no "normal" workload gets deployed on the controller nodes. So the controller nodes are now running the normal control plane processes like `kube-apiserver` and so on and additionally `kube-proxy` and `kubelet`. And since I'm using `Cilium` for all K8s related networking stuff that is running too on that nodes. It's not perfect and there are other options out there but none of them were that appealing to me.
